### ISSUE TITLE
Change 1st batch/3 arg to function pointer

### ISSUE THREAD

**Issue by mgwidmann** [absinthe-graphql/absinthe#407]
Currently the batch function is a little rough to use since the first argument takes `{module, :function_name} | {module, :function_name, term}`. It would be nicer if instead it also supported `&SomeModule.some_fun/2`.
That would change
```elixir
      resolve fn post, _, _ ->
        batch({__MODULE__, :users_by_id}, post.author_id, fn batch_results ->
          {:ok, Map.get(batch_results, post.author_id)}
        end)
      end
```

To

```elixir
      resolve fn post, _, _ ->
        batch(&__MODULE__.users_by_id/2, post.author_id, fn batch_results ->
          {:ok, Map.get(batch_results, post.author_id)}
        end)
      end
```

Is there any reason why this isnt already supported?

**Comment by benwilson512** [comment]
Right so the problem is, if you accept that form then you would expect to be able to do &foo(&1, &2, :bar) or any of the other anonymous function forms. Those however won't work because anonymous function equality is always false, preventing any actual batching from happening.

**Comment by benwilson512** [comment]
Was on mobile at the time, so I'll elaborate a bit more here. The first argument to the `batch/3` is a crucial part of grouping the data that belongs in a batch. Naturally, only things that have the same batch function should be grouped. The problem is, "sameness" for anonymous functions breaks down pretty quickly.

Was on mobile at the time, so I'll elaborate a bit more here. The first argument to the `batch/3` is a crucial part of grouping the data that belongs in a batch. Naturally, only things that have the same batch function should be grouped. The problem is, "sameness" for anonymous functions breaks down pretty quickly.

```elixir
> f = &IO.puts/1
&IO.puts/1
> g = &IO.puts/1
&IO.puts/1
> f == g
true
```
```elixir
iex(18)> f = &IO.puts(:stderr, &1)
#Function<6.99386804/1 in :erl_eval.expr/5>
iex(19)> g = &IO.puts(:stderr, &1)
#Function<6.99386804/1 in :erl_eval.expr/5>
iex(20)> f == g
false
```
```elixir
> f = fn str -> IO.puts(str) end 
#Function<6.99386804/1 in :erl_eval.expr/5>
> g = fn str -> IO.puts(str) end
#Function<6.99386804/1 in :erl_eval.expr/5>
> f == g
false
```

In short, we could only allow this if we restricted people to using _exactly_ the zero closure anonymous function form, and since 99% of the time this distinction doesn't matter, I'm concerned we'd be setting our users up for confusion, and ourselves up to answer this question over and over.

**Comment by benwilson512** [comment]
FWIW I _do_ think that the batch API sorta sucks, and I've been thinking about alternatives. They just aren't super high on the priority list right now.

**Comment by mgwidmann** [comment]
Yeah I see. I agree though its a little rough around the edges... If you look at how shopify's graphql-batch gem works, it may help inspire a better syntax. It uses syntax like this.

```ruby
# First you create a class that extends the batch loader provided
class RecordLoader < GraphQL::Batch::Loader
  def initialize(model)
    @model = model
  end

  # Here is where the current `batch/3`'s first argument is defined, it is implicitly executed
  def perform(ids)
    @model.where(id: ids).each { |record| fulfill(record.id, record) }
    ids.each { |id| fulfill(id, nil) unless fulfilled?(id) }
  end
end

# Then when you want to fetch an object you use it like this:
resolve -> (obj, args, context) { RecordLoader.for(Product).load(args["id"]) }
```

They are calling `fulfill(key, value)` for every entry they want to fill out, that basically is the same as the current `batch/3`'s last parameter. The benefit to this loader is that it can be reused for any belongs_to relationship. You can then create one for each kind of relationship based upon where the foreign key is.

With a few modifications to the above `RecordLoader` you can make a has many loader or a many to many loader. This makes those loaders reusable.

Something similar in Elixir can be done. Perhaps something like this might be more readable?

```elixir
defmodule RecordLoader do
  use Absinthe.Batch
  import Ecto.Query
  alias MyApp.Repo

  batch do
    # The meat of the whole thing
    perform fn  [model], ids -> # first arg comes from below `for/1` call
      from(m in model, where: m.id in ^ids) 
      |> Repo.all
      |> Enum.group_by(&:id) # Returns a map of `%{ id => struct}`
    end

    # How to get the data out, this could be optional
    resolve fn map, id ->
      map[id]
    end
  end
end

# Then you'd use it in your resolve like
resolve fn _, %{id: id}, _ ->
  # This will identify a loader's signature, meaning the uniqueness of it is key to grouping properly
  RecordLoader.for([Product])
  # Queues up this argument to the loader identified by `for/1`
  |> RecordLoader.load(id)
end
```
